### PR TITLE
Create an inbox note for Stripe Link set-up

### LINKED ISSUE #4844: StripeLink: Inbox notification
### Description

A new inbox notification has to be added in admin WC -> home

Title: Increase conversion at checkout
Copy: Reduce cart abandonment and create a frictionless checkout experience with Link by Stripe. Link autofills your customer’s payment and shipping details so they can check out in just six seconds with the Link optimized experience. That’s 9x faster than shoppers who don’t use Link. Link increases conversion rates by over 7% for logged-in Link customers.
[CTA: Set up now](https://woocommerce.com/document/payments/woocommerce-payments-stripe-link/)
→ links to the docs above.

### PR DESCRIPTION
**Pull request by mordeth**
Fixes #4844

#### Changes proposed in this Pull Request
This PR adds an inbox note when:
- Stripe Link is not enabled;
- Stripe is an available payment method ( Country US );
- UPE is enabled;

Screenshot 2022-10-17 at 12 36 53

Screenshot 2022-10-17 at 12 37 24

#### Testing instructions

*Case 1*
- Verify Stripe Link payment method is available
- Disable Stripe Link payment method if already enabled
- Verify the note is added to the Inbox section on `WooCommerce` -> `Home`

*Case 2*
- Disable UPE by updating `_wcpay_feature_upe` option to `0`
- Deactivate and Activate WCPay
- Verify the note is not added to the Inbox section on `WooCommerce` -> `Home`

*Case 3*
- Enable UPE by updating `_wcpay_feature_upe` to `1`
- Enable Stripe Link from 
- Deactivate and Activate WCPay
- Verify the note is not added to the Inbox section on `WooCommerce` -> `Home`

*Case 4*
- Connect to non-US merchant account
- Deactivate and Activate WCPay
- Verify the note is not added to the Inbox section on `WooCommerce` -> `Home`

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	/**
+	 * Checks if a note can and should be added.
+	 *
+	 * @return bool
+	 */
+	public static function should_display_note() {
+		$payment_method_statuses = self::$gateway->get_upe_enabled_payment_method_statuses();
+		$status                  = $payment_method_statuses['link_payments']['status'] ?? null;
+
+		if ( 'unrequested' === $status ) {
```
Comment (kalessil) [inline includes/notes/class-wc-payments-notes-set-up-stripelink.php]: IIRC, when we complete migration from express to standard accounts, it'll be enabled by default.
So we should follow the agree criteria: p1666002659690399/1666000295.406309-slack-C02C00HCJ8J

Comment (kalessil) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Additionally, Link is part of the UPE, and we need ensure that account is connected and UPE available.
Some clues: `\WC_Payments_Notes_Additional_Payment_Methods::get_note` to discover further.

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	 */
+	private static $gateway;
+
+	/**
+	 * Checks if a note can and should be added.
+	 *
+	 * @return bool
+	 */
+	public static function should_display_note() {
+		$payment_method_statuses = self::$gateway->get_upe_enabled_payment_method_statuses();
```
Comment (kalessil) [inline includes/notes/class-wc-payments-notes-set-up-stripelink.php]: And what happens if the gateway is not yet injected?

**Review by kalessil — CHANGES_REQUESTED** [review]
I left a couple of comments + would like to remind about adding some tests for this (see tests of `WC_Payments_Notes_Set_Up_Refund_Policy` for inspiration).

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php (hunk 1 above)
Comment (mordeth) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Not applicable anymore.

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php (hunk 2 above)
Comment (mordeth) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Not applicable anymore.

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	public static function get_note() {
+		if ( ! self::should_display_note() ) {
+			return;
+		}
+
+		$note_class = WC_Payment_Woo_Compat_Utils::get_note_class();
+		$note       = new $note_class();
+
+		$note->set_title( __( 'Increase conversion at checkout', 'woocommerce-payments' ) );
+		$note->set_content( __( 'Reduce cart abandonment and create a frictionless checkout experience with Link by Stripe. Link autofills your customer’s payment and shipping details so they can check out in just six seconds with the Link optimized experience. That’s 9x faster than shoppers who don’t use Link. Link increases conversion rates by over 7% for logged-in Link customers.', 'woocommerce-payments' ) );
```
Comment (mgascam) [inline includes/notes/class-wc-payments-notes-set-up-stripelink.php:78]: This line seems too long, can we break it for better readability?

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	 * @var WC_Payment_Gateway_WCPay
+	 */
+	private static $gateway;
+
+	/**
+	 * Checks if a note can and should be added.
+	 *
+	 * @return bool
+	 */
+	public static function should_display_note() {
```
Comment (mgascam) [inline includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Can we add return type here?
```suggestion
	public static function should_display_note():bool {
```

**Review by mgascam — APPROVED** [review]
LGTM! I followed testing cases and all test well.

I left a couple of minor comments about type hints (added one comment but I think there's also room for more) and formatting.

File: tests/unit/notes/test-class-wc-payments-notes-set-up-stripelink.php [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+		$this->mock_action_scheduler_service = $this->getMockBuilder( 'WC_Payments_Action_Scheduler_Service' )
+			->disableOriginalConstructor()
+			->getMock();
+
+		$this->mock_rate_limiter = $this->createMock( Session_Rate_Limiter::class );
+
+		$this->order_service = new WC_Payments_Order_Service( $this->mock_api_client );
+
+		$this->mock_wcpay_gateway = $this->getMockBuilder( '\WC_Payment_Gateway_WCPay' )
```
Comment (kalessil) [inline tests/unit/notes/test-class-wc-payments-notes-set-up-stripelink.php:24]: Looking into the test case, we set some expectations on `mock_wcpay_gateway` only. Could we omit to introduce unneeded properties and relocate the mocks construction right into constructor arguments, perhaps?

**Review by dpaun1985 — APPROVED** [review]
LGTM!
Tested and works as expected.

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		// If UPE is not enabled, skip.
+		if ( ! \WC_Payments_Features::is_upe_enabled() ) {
+			return false;
+		}
+
+		// Check if Link payment is available.
+		$available_upe_payment_methods = self::$gateway->get_upe_available_payment_methods();
+
```
Comment (kalessil) [inline includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Can we drop this empty line, please?

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+		if ( ! in_array( Link_Payment_Method::PAYMENT_METHOD_STRIPE_ID, $available_upe_payment_methods, true ) ) {
+			return false;
+		}
+
+		// Retrieve enabled payment methods at checkout.
+		$enabled_payment_methods = self::$gateway->get_payment_method_ids_enabled_at_checkout( null, true );
+
```
Comment (kalessil) [inline includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Can we drop this empty line, please?

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+		// If card payment method is not enabled, skip.
+		if ( ! in_array( CC_Payment_Method::PAYMENT_METHOD_STRIPE_ID, $enabled_payment_methods, true ) ) {
```
Comment (kalessil) [inline includes/notes/class-wc-payments-notes-set-up-stripelink.php]: To keep related pieces together, can we merge this and next if-statements into one, please?

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php (hunk 3 above)
Comment (mordeth) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php:78]: Originally I tried to separate it as a multi-line string, but it is not a valid param for `__()` method. Should keep it like this.

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php (hunk 4 above)
Comment (mordeth) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Done.

File: tests/unit/notes/test-class-wc-payments-notes-set-up-stripelink.php (hunk 5 above)
Comment (mordeth) [reply tests/unit/notes/test-class-wc-payments-notes-set-up-stripelink.php:24]: Makes sense. I have disabled the constructor and used a single mock.

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php (hunk 6 above)
Comment (mordeth) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Done

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php (hunk 8 above)
Comment (mordeth) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Done

File: includes/notes/class-wc-payments-notes-set-up-stripelink.php (hunk 7 above)
Comment (mordeth) [reply includes/notes/class-wc-payments-notes-set-up-stripelink.php]: Done

**Review by kalessil — APPROVED** [review]
LGTM!

**Pull request merged**